File: 801→900/837. New 21 Game/solution.js
/**
 * @param {number} n
 * @param {number} k
 * @param {number} maxPts
 * @return {number}
 */
var new21Game = function (n, k, maxPts) {
  if (k === 0 || n >= k + maxPts) return 1;
  const dp = new Array(n + 1).fill(0);
  dp[0] = 1;
  let sum = 1;
  for (let i = 1; i <= n; ++i) {
    dp[i] = sum / maxPts;
    if (i < k) sum += dp[i];
    if (i - maxPts >= 0) sum -= dp[i - maxPts];
  }
  let res = 0;
  for (let i = k; i <= n; ++i) res += dp[i];
  return res;
};

File: 801→900/837. New 21 Game/solution.cpp
class Solution {
public:
    double new21Game(int n, int k, int maxPts) {
        vector<double> dp(n + 1);

        dp[0] = 1;

        double s = k > 0 ? 1 : 0;

        for (int i = 1; i <= n; i++) {
            dp[i] = s / maxPts;
            if (i < k)
                s += dp[i];

            if (i - maxPts >= 0 && i - maxPts < k)
                s -= dp[i - maxPts];

            // for (int j = 1; j <= maxPts; j++) {
            //     if (i - j >= 0 && i - j < k) {
            //         dp[i] += dp[i - j] / maxPts;
            //     }
            // }
        }
        return accumulate(dp.begin() + k, dp.end(), 0.0);
    }
};